### PR TITLE
Updating typings config to reflect actual callback parameters

### PR DESCRIPTION
**Pull request by skyflyer**
## Description

This is a typescript typings file with corrected definitions for properties and callbacks on `Datetime` object.

## Motivation and Context

This PR should be merged to avoid Typescript compilation errors that result in using the callbacks with wrong callback parameters.

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] No javascript code has been changed

### CODE REVIEW

File: react-datetime.d.ts [hunk 1]
```diff
@@ -50,17 +50,17 @@ declare module ReactDatetime {
      only parameter, if the date in the input is valid. If it isn't, the value
      of the input (a string) is returned.
      */
-    onChange?:(x: string) => void;
+    onChange?:(x: string|any) => void;
```
Comment (simeg) [inline react-datetime.d.ts]: Please add space between `onChange?:` and `(x: string ...)` to comply with styling of the rest of the file.

File: react-datetime.d.ts [hunk 2]
```diff
@@ -50,17 +50,17 @@ declare module ReactDatetime {
...
      */
-    onFocus?: (e : FocusEvent) => void;
+    onFocus?: () => void;
     /*
      Callback trigger for when the user clicks outside of the input, simulating a regular onBlur.
      The callback receives the selected moment object as only parameter, if the date in the
      input is valid. If it isn't, the value of the input (a string) is returned.
      */
-    onBlurs?: (e : FocusEvent) => void;
+    onBlur?: (e : string|any) => void;
```
Comment (simeg) [inline react-datetime.d.ts]: Nice find with the typo!

File: react-datetime.d.ts (hunk 1 above)
Comment (simeg) [inline react-datetime.d.ts]: Doesn't `onChange` accept a `function` and not `string` or `any` as input?

File: react-datetime.d.ts [hunk 3]
```diff
@@ -50,17 +50,17 @@ declare module ReactDatetime {
      only parameter, if the date in the input is valid. If it isn't, the value
      of the input (a string) is returned.
      */
-    onChange?:(x: string) => void;
+    onChange?:(x: string|any) => void;
     /*
      Callback trigger for when the user opens the datepicker.
      */
-    onFocus?: (e : FocusEvent) => void;
+    onFocus?: () => void;
```
Comment (simeg) [inline react-datetime.d.ts]: Doesn't `onFocus` accept a `function` and not `string` or `any` as input?

File: react-datetime.d.ts (hunk 2 above)
Comment (simeg) [inline react-datetime.d.ts]: Doesn't `onBlur` accept a `function` and not `string` or `any` as input?

File: react-datetime.d.ts (hunk 1 above)
Comment (simeg) [inline react-datetime.d.ts]: Doesn't this actually return something according to the docs? 

> The callback receives the selected moment object as only parameter, if the date in the input is valid. If it isn't, the value of the input (a string) is returned.

Or is the documentation a bit confusing? First it mentions the parameter, and then it says "If it isn't - return string". Maybe I'm just misunderstanding something here?

File: react-datetime.d.ts (hunk 2 above)
Comment (simeg) [inline react-datetime.d.ts]: Same thing about `onBlur` as I mentioned above with `onChange`. I'm a bit confused, I'd love an explanation if you have one :)

File: react-datetime.d.ts [hunk 4]
```diff
@@ -77,27 +77,27 @@ declare module ReactDatetime {
      Define the dates that can be selected. The function receives (currentDate, selectedDate)
      and should return a true or false whether the currentDate is valid or not. See selectable dates.
      */
-    isValidDate?: (x: string) => void;
+    isValidDate?: (x: any, y: any) => boolean;
```
Comment (simeg) [inline react-datetime.d.ts]: Are `x` and `y` very good names for the parameters? In this case we know that `x` represents the `currentDate`, shouldn't we use that as the name of the parameter then instead? This should apply to all definitions in this file then.

File: react-datetime.d.ts [hunk 5]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
      When true, once the day has been selected, the react-datetime will be automatically closed.
      */
     closeOnSelect?: boolean;
+
```
Comment (simeg) [inline react-datetime.d.ts]: Remove this empty line, please.

File: react-datetime.d.ts [hunk 6]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
...
+    /*
+    Allow to add some constraints to the time selector. It accepts an object with the format {hours:{ min: 9, max: 15, step:2}}
```
Comment (simeg) [inline react-datetime.d.ts]: It seems like the styling of this block isn't correct, if you compare it with the other comment blocks. It should be pushed to the right a bit more. Also, please break the line manually so it doesn't have to break on it's own (look how long the lines are in other parts of this file). :)

File: react-datetime.d.ts [hunk 7]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
...
+    so the hours can't be lower than 9 or higher than 15, and it will change adding or subtracting 2 hours everytime the 
+    buttons are clicked. The constraints can be added to the hours, minutes, seconds and milliseconds.
+    */
+    timeConstraints: any; 
+    /*
+    When true, keep the picker open when click event is triggered outside of component. When false, close it.
+    */
+    disableOnClickOutside?:	boolean;
```
Comment (simeg) [inline react-datetime.d.ts]: There's some extra space between `disableOnClickOutside?:` and `boolean;` that can be removed.

File: react-datetime.d.ts [hunk 8]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
...
     closeOnSelect?: boolean;
+
+    /*
+    Allow to add some constraints to the time selector. It accepts an object with the format {hours:{ min: 9, max: 15, step:2}}
+    so the hours can't be lower than 9 or higher than 15, and it will change adding or subtracting 2 hours everytime the 
+    buttons are clicked. The constraints can be added to the hours, minutes, seconds and milliseconds.
+    */
+    timeConstraints: any; 
+    /*
+    When true, keep the picker open when click event is triggered outside of component. When false, close it.
```
Comment (simeg) [inline react-datetime.d.ts]: Same with this comment block, should be indented properly.

File: react-datetime.d.ts [hunk 9]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
...
+    */
+    disableOnClickOutside?:	boolean;
+
```
Comment (simeg) [inline react-datetime.d.ts]: This line can be removed as well.

File: react-datetime.d.ts [hunk 10]
```diff
@@ -106,6 +106,18 @@ declare module ReactDatetime {
      When true, once the day has been selected, the react-datetime will be automatically closed.
      */
     closeOnSelect?: boolean;
+
+    /*
+    Allow to add some constraints to the time selector. It accepts an object with the format {hours:{ min: 9, max: 15, step:2}}
+    so the hours can't be lower than 9 or higher than 15, and it will change adding or subtracting 2 hours everytime the 
+    buttons are clicked. The constraints can be added to the hours, minutes, seconds and milliseconds.
+    */
+    timeConstraints: any;
```
Comment (simeg) [inline react-datetime.d.ts]: The `any` should be replaced with an `object`.

**Comment by skyflyer** [comment]
Simon, I appreciate your comments. As I was just fixing them to have the right signature, I didn't think about actually improving the signatures of each definition, as `x` is as obvious to me as `date`, but your comments are spot on! Having said that, let me jump right in and address your comments

File: react-datetime.d.ts (hunk 1 above)
Comment (skyflyer) [inline react-datetime.d.ts]: I believe the documentation is phrased in a way that leads to confusion. If we look through source code, we see that the result of `onChange` is never used. The documentation string is taken from the root `README.md` file. I've amended that part of the configuration as well. Ref: `DateTime.js:184`.

**Comment by skyflyer** [comment]
Simon:

* renderDay: I've changed the signature based on the doc snippet
```
    renderDay: function( props, currentDate, selectedDate ){
        return <td {...props}>{ '0' + currentDate.date() }</td>;
    },
```
* renderMonth: similar to renderDay above, but different signature
* renderYear: similar to above

File: README.md [hunk 11]
```diff
@@ -42,9 +42,9 @@ render: function() {
...
 | **locale** | `string` | `null` | Manually set the locale for the react-datetime instance. Moment.js locale needs to be loaded to be used, see [i18n docs](#i18n).
-| **onChange** | `function` | empty `function` | Callback trigger when the date changes. The callback receives the selected `moment` object as only parameter, if the date in the input is valid. If it isn't, the value of the input (a `string`) is returned. |
-| **onFocus** | `function` | empty `function` | Callback trigger for when the user opens the datepicker. |
-| **onBlur** | `function` | empty `function` | Callback trigger for when the user clicks outside of the input, simulating a regular onBlur. The callback receives the selected `moment` object as only parameter, if the date in the input is valid. If it isn't, the value of the input (a `string`) is returned. |
+| **onChange** | function | empty function | Callback trigger when the date changes. The callback receives the selected `moment` object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback receives the value of the input (a string). |
```
Comment (simeg) [inline README.md]: Please stay consistent with the formatting of the types, `function` is a valid type so don't remove the backticks.

File: README.md [hunk 12]
```diff
@@ -42,9 +42,9 @@ render: function() {
...
+| **onFocus** | function | empty function | Callback trigger for when the user opens the datepicker. |
```
Comment (simeg) [inline README.md]: (Same here) Please stay consistent with the formatting of the types, `function` is a valid type so don't remove the backticks.

File: README.md [hunk 13]
```diff
@@ -57,6 +57,7 @@ render: function() {
 | **closeOnTab** | `boolean` | `true` | When `true` and the input is focused, pressing the `tab` key will close the datepicker.
 | **timeConstraints** | `object` | `null` | Add some constraints to the timepicker. It accepts an `object` with the format `{ hours: { min: 9, max: 15, step: 2 }}`, this example means the hours can't be lower than `9` and higher than `15`, and it will change adding or subtracting `2` hours everytime the buttons are clicked. The constraints can be added to the `hours`, `minutes`, `seconds` and `milliseconds`.
 | **disableOnClickOutside** | `boolean` | `false` | When `true`, keep the datepicker open when click event is triggered outside of component. When `false`, close it.
+>>>>>>> upstream/master
```
Comment (simeg) [inline README.md]: This should be removed.

**Comment by simeg** [comment]
Hey @skyflyer. You're adding a lot of nice things, awesome! I just had some formatting comments, please look them over and I'll merge your changes :) (I still haven't figured out this new review thingy)

**Pull request merged**